src/main/java/net/imagej/ops/OpListing.java

```diff
@@ -0,0 +1,348 @@
+	public List<ModuleItem<?>> inputsFor(final ModuleInfo info) {
+		return IntStream.range(0, inputNames.size()) //
+			.mapToObj(i -> new DefaultMutableModuleItem<>( //
+				info, //
+				inputNames.get(i), //
+				Types.raw(inputTypes.get(i)))) //
+			.collect(Collectors.toList());
+	}
+
+	/**
+	 * @return the {@link Type}s of each input, in order, required by the listing.
+	 */
+	public List<Type> getInputTypes() {
+		return inputTypes;
+	}
+
+	/**
+	 * @return the names of each input, in order, required by the listing.
+	 */
+	public List<String> getInputNames() {
+		return inputNames;
+	}
+
+	/**
+	 * Creates a set of {@link ModuleItem}s for <b>each</b> output described by
+	 * this {@link OpListing}.
+	 *
+	 * @param info the {@link ModuleInfo} to which the resulting
+	 *          {@link ModuleItem}s will belong.
+	 * @return {@link ModuleItem}s describing the outputs of this
+	 *         {@link OpListing}
+	 */
+	public List<ModuleItem<?>> outputsFor(final ModuleInfo info) {
+		if (returnName.isPresent() && returnType.isPresent()) {
+			final String name = returnName.get();
+			final Class<?> type = Types.raw(returnType.get());
+			final MutableModuleItem<?> outItem = //
+				new DefaultMutableModuleItem<>(info, name, type);
+			outItem.setIOType(ItemIO.OUTPUT);
+			return Collections.singletonList(outItem);
+		}
+		return Collections.emptyList();
+	}
+
+	/**
+	 * @return the {@link Type}s of each output, in order, required by the
+	 *         listing.
+	 */
+	public Optional<Type> getReturnType() {
+		return returnType;
+	}
+
+	/**
+	 * @return the names of each output, in order, required by the listing.
+	 */
+	public Optional<String> getReturnName() {
+		return returnName;
+	}
+
+	@Override
+	public boolean equals(final Object obj) {
+		if (!(obj instanceof OpListing)) return false;
+		final OpListing that = (OpListing) obj;
+
+		return Objects.equals(getName(), that.getName()) && //
+			Objects.equals(getFunctionalType(), that.getFunctionalType()) && //
+			Objects.equals(getInputTypes(), that.getInputTypes()) && //
+			Objects.equals(getInputNames(), that.getInputNames()) && //
+			Objects.equals(getReturnType(), that.getReturnType()) && //
+			Objects.equals(getReturnName(), that.getReturnName());
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash( //
+			getName(), //
+			getFunctionalType(), //
+			getInputTypes(), //
+			getInputNames(), //
+			getReturnType(), //
+			getReturnName() //
+		);
+	}
+
+	@Override
+	public String toString() {
+		final StringBuilder sb = new StringBuilder();
+		// name
+		sb.append(getName());
+		// inputs
+		sb.append("(").append(paramList(inputNames, inputTypes)).append(")");
+		// (pure) output
+		if (returnName.isPresent() && returnType.isPresent()) {
+			final String returnString = shortParam(returnName.get(), returnType
+				.get());
+			sb.append(" -> ").append(returnString);
+		}
+		return sb.toString();
+	}
+
+	public OpListing reduce(final Function<Type, Type> typeReducer) {
```

Comment:
I don't have a clear vision for how this will be used, externally. I understand that it's an attempt to make it possible to generalize the `OpListing` logic across more types than those hardcoded by `SIG_SIMPLIFICATIONS`, but in practice, there needs to be a way for napari-imagej to tell the `OpService` that it wants to use some alternative simplifications all the time when searching, right? This seems like it should be global state to the `OpService`. And then the `reduce` guts here could stay private.

A more general way to deal with this, without custom functions, would be for the `typeReducer` to lean on the `ConvertService`. If a type can be converted to e.g. `RAI`, then it is reduced to the string `"image"`. If it can be converted to `Number`, then it is reduced to the string `"number"`. Or maybe we don't want to literally use strings as pseudo-types, we can still pick some canonical Java type for each of image and number perhaps, since I imagine this would be important for the `OpListingModule`.

Reply:
I think the `ConvertService` will be over-eager here. We don't have an `Object` here, only `Type`s. Thus we'll have `String` inputs/outputs be reduced to `Img`s, and we'd have to increase complexity to make that work. I'd say we leave it for now, unless you have more thoughts/ideas.